apps/rabbitmq_prelaunch/src/rabbit_boot_state_systemd.erl

```diff
@@ -0,0 +1,160 @@
+%%%-------------------------------------------------------------------
+%% The contents of this file are subject to the Mozilla Public License
+%% Version 1.1 (the "License"); you may not use this file except in
+%% compliance with the License. You may obtain a copy of the License
+%% at https://www.mozilla.org/MPL/
+%%
+%% Software distributed under the License is distributed on an "AS IS"
+%% basis, WITHOUT WARRANTY OF ANY KIND, either express or implied. See
+%% the License for the specific language governing rights and
+%% limitations under the License.
+%%
+%% The Original Code is RabbitMQ.
+%%
+%% The Initial Developer of the Original Code is GoPivotal, Inc.
+%% Copyright (c) 2007-2020 Pivotal Software, Inc.  All rights reserved.
+%%
+
+-module(rabbit_boot_state_systemd).
+
+-behavior(gen_server).
+
+-export([start_link/0]).
+
+-export([init/1,
+         handle_call/3,
+         handle_cast/2,
+         handle_info/2,
+         terminate/2,
+         code_change/3]).
+
+-record(state, {sd_notify_module,
+                socket}).
+
+
+start_link() ->
+    gen_server:start_link({local, ?MODULE}, ?MODULE, [], []).
+
+init([]) ->
+    case code:load_file(sd_notify) of
+        {module, sd_notify} ->
+            {ok, #state{sd_notify_module = sd_notify}};
+        {error, _} ->
+            case rabbit_prelaunch:get_context() of
+                #{systemd_notify_socket := Socket} when Socket =/= undefined ->
+                    {ok, #state{socket = Socket}};
+                _ ->
+                    ignore
+            end
+    end.
+
+handle_call(_Request, _From, State) ->
+    {noreply, State}.
+
+handle_cast({notify_boot_state, ready}, #{sd_notify_module := SDNotify} = State) ->
+    rabbit_log_prelaunch:debug("notifying systemd of readiness via native module"),
+    sd_notify_legacy(SDNotify),
+    {noreply, State};
+handle_cast({notify_boot_state, ready}, #{socket := Socket} = State) ->
+    rabbit_log_prelaunch:debug("notifying systemd of readiness via socat"),
+    sd_notify_socat(Socket),
+    {noreply, State};
+handle_cast({notify_boot_state, _BootState}, State) ->
+    {noreply, State}.
+
+handle_info(Msg, State) ->
+    io:format(standard_error, "Unexpected message: ~p~n",[Msg]),
```

Comment:
I suppose this is a debug message. We almost never write directly to stdout/stderr. You can log a debug message for this one if you want, but you need to put more context because the message won't tell which module/process issued it.

Reply:
the original systemd code logs in this fashion, which I agree isn't ideal. I believe this would only occur in the case that the boot state listener contract is not fulfilled, and this is more a developer than a runtime error. Is it common practice to raise an exception in such a situation?

Reply:
I guess the other messages can use the prelaunch logger, as they won't occur until that logger is ready.

Reply:
Yes. You should use pattern matching if possible and let the code crash if there is an invalid input. This module is internal (not exposed to user input), so it makes sense to crash when the input is invalid.

Here, you have to implement `handle_info()` so you can't really use pattern matching here because we expect no message here.

Reply:
I think under the circumstances it's best to just leave the improved log message. We could `{stop, unexpected_message, State}` but if I'm not mistaken, that would take the whole broker down, and since these listeners feel of lower precedence, I would avoid that.

Reply:
For now, you can change the message to something logged (`rabbit_log:debug(...)` or `rabbit_log:warning(...)`, as you wish).

Reply:
since handle_info is optional, and it wasn't clear to me if this should be a `rabbit_log_prelaunch:debug` or `rabbit_log:debug`, I decided to just remove the function.